KAFKA-6042: Avoid deadlock between two groups with delayed operations

rajinisivaram opened the pull request.

rajinisivaram committed:
KAFKA-6042: Avoid deadlock between two groups with delayed operations

rajinisivaram commented:
@ijuma @hachikuji Can you review please? I was trying to avoid changing group/txnMetadata locking last time,  but that was a mistake. This goes back to the previous way of overriding locks, but with `ReentrantLock`.

rajinisivaram committed:
Do appendRecords without txnMetadata lock in TransactionStateManager

hachikuji commented on core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala:
```diff
@@ -586,24 +587,29 @@ class TransactionStateManager(brokerId: Int,
 
-          metadata synchronized {
+          val append: Boolean =
+          metadata.inLock {
```
nit: move this to previous line?

hachikuji reviewed (approved):
Left a nit, but LGTM overall. It might help to find somewhere to add some notes on the locking strategy both for the `GroupCoordinator` and the `TransactionCoordinator`.

hachikuji commented:
@rajinisivaram Hmm.. Do we need to update `DelayedHeartbeat` also to use the group lock as the operation lock.

rajinisivaram committed:
Address review comments, add comments

rajinisivaram commented:
@hachikuji I updated `DelayedHeartbeat` to use group lock and `DelayedTxnMarker` to use state lock. Also added comments in `GroupCoordinator` and `TransactionStateManager`. Can you take another look? Thank you!

hachikuji commented:
The updates look good to me. I'll wait for @guozhangwang to take a second look before merging.

guozhangwang reviewed (commented):
I have a meta question for my own understanding: the original reported deadlock in KAFKA-6042 seems between two offset commit request for two groups, one holding group A and tries to unblock for the other parked delayed producer for group B, and the other vice versa. This PR seems to be focusing on the transaction metadata manager instead. Does this actually fix the reported issue?

Otherwise I do not have any comments. Nice replacing `synchronized` with lock.

rajinisivaram commented:
@guozhangwang The PR fixes deadlocks in both group coordinator and transaction manager. For groups, the group lock is now used as the delayed operation lock . The deadlock reported was exactly as you described. Now, when offset commit  request for groupA tries to complete groupB, the completion is performed if and only if it is able to acquire groupB lock (using `tryLock`). And similarly for groupB completing groupA. So we may acquire two group locks in any order, but the cycle is broken by `tryLock`, preventing deadlock.

rajinisivaram commented:
please retest this

rajinisivaram commented:
retest this please

guozhangwang commented:
> And similarly for groupB completing groupA. So we may acquire two group locks in any order, but the cycle is broken by tryLock, preventing deadlock.

Thanks for the explanation. LGTM.

guozhangwang commented:
The jenkins failures seem un-related. I'm going to merge to trunk and 1.0.

guozhangwang commented:
Merged to 1.0, and cherry-picked to trunk.

ijuma commented on core/src/main/scala/kafka/server/DelayedOperation.scala:
```diff
@@ -43,11 +43,12 @@ import scala.collection.mutable.ListBuffer
  */
-abstract class DelayedOperation(override val delayMs: Long) extends TimerTask with Logging {
+abstract class DelayedOperation(override val delayMs: Long,
+    lockOpt: Option[Lock] = None) extends TimerTask with Logging {
```
I suggest reconsidering all default arguments used for locks. Since it's crucial to get them right, it's generally safer to force callers to choose. My general recommendation is to avoid default arguments whenever correctness is involved.

boweite commented:
@ijuma @hachikuji @guozhangwang Hi, I wonder if this patch can completely solve this problem. Because I found the order of the parameters passed from replicaManager.appendRecords is wrong. So the group.lock is not passed to delayedProduce, so there is also the risk of deadlock. Please check it.
